### PR TITLE
add duration if the query is canceled when it needs to be logged

### PR DESCRIPTION
**Pull request by SmartKeyerror**
Add the duration time into "canceling statement due to user request" to help the user
debug when canceling the query.

The log message will be

```
2023-10-10 03:32:55.325528 UTC,"smart","postgres",p463714,th1746153216,"[local]",,2023-10-10 03:32:01 UTC,
0,con32,cmd9,seg-1,,,,sx1,"ERROR","57014","canceling statement due to user request, duration:4256.617",,,,,,
"select * from t, t as t1, t as t2;",0,,"postgres.c",4053,
```

Not change the log schema.

### CODE REVIEW

**Review by QingMa-gp — APPROVED** [review]
LGTM.

**Comment by SmartKeyerror** [comment]
Dev pipeline test cases are all passed, https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/lzl-log-cancel-duration?group=all, merge it.

File: src/backend/tcop/postgres.c [hunk 1]
```diff
@@ -4038,9 +4038,25 @@ ProcessInterrupts(const char* filename, int lineno)
 						(errcode(ERRCODE_GP_OPERATION_CANCELED),
 						 errmsg("canceling MPP operation%s", cancel_msg_str.data)));
 			else
-				ereport(ERROR,
-						(errcode(ERRCODE_QUERY_CANCELED),
-						 errmsg("canceling statement due to user request%s", cancel_msg_str.data)));
+			{
+				char		msec_str[32];
+
+				switch (check_log_duration(msec_str, false))
```
Comment (ashwinstar) [inline src/backend/tcop/postgres.c:4044]: usage of `check_log_duration()` to get the duration value populated seems fine. Though using `log_min_duration_statement` as side effect to control when to display duration in this cancellation log message seems confusing. As by default `log_min_duration_statement` is set to -1 (disabled) and just to get cancellation queries duration will have to set it to some reasonable value which will also enable regular query stmts logging, that seems not a desired behavior.

Full requirement of how this value will be used and what exact purpose it serves - isn't clear to me. Though feels it has different usecase compared to regular `log_duration` and `log_min_duration_statement` aspects. Hence, shouldn't be combined together. Maybe for cancelling if really wish to provide the duration then just do independent of the return value from `check_log_duration()`.

File: src/backend/tcop/postgres.c [hunk 2]
```diff
@@ -4038,9 +4038,25 @@ ProcessInterrupts(const char* filename, int lineno)
...
+					case 0:
+						ereport(ERROR,
+								(errcode(ERRCODE_QUERY_CANCELED),
+										errmsg("canceling statement due to user request%s", cancel_msg_str.data)));
+						break;
+					case 1:
+					case 2:
+						ereport(ERROR,
+								(errcode(ERRCODE_QUERY_CANCELED),
+										errmsg("canceling statement due to user request%s, duration:%s",
```
Comment (ashwinstar) [inline src/backend/tcop/postgres.c:4055]: just `duration` makes it harder to interpret - is it duration for query to execute till cancellation or its duration for cancelling the query since cancellation request is received. Maybe we can reword it to explicitly say "query elapsed time:"

File: src/backend/tcop/postgres.c (hunk 1 above)
Comment (ashwinstar) [reply src/backend/tcop/postgres.c:4044]: Based on offline discussions - requirement is get the query duration for successful queries which currently can be done by enabling `log_duration`. and for cancelled queries. Currently, no easy way to know how long cancelled query was executing. Hence using `log_duration` or `log_min_duration_statement` to control the behavior of displaying duration for query both for success and cancelled case seems perfect solution and meeting the requirements.

Plus, usage of the "duration" word then also keeps it consistent between both the usages.
